### PR TITLE
Support for pinned messages

### LINKED ISSUE #57: Notification - Pinned Messages
Hi,

thank you so much for Fernschreiber1

I encountered the issue that when I muted chats I get a notification that the pinned message was updated anyways. Also in the notification and in the chat it is written: "Unsupported message: PinMessage". 

It is kind of annoying. How can I disable that? I do not find any option in the pully menu. 

Is this a bug?

### PR DESCRIPTION
**Pull request by Wunderfitz**
OK, that escalated quickly. Supporting pinned messages is far more work than expected, especially if you combine it with the introduction of a nice overlay that has the ability to display messages. This overlay is now used for pinned messages and in-reply-to messages. Simply give it a try!

You can now also pin messages yourself if you have proper authorizations in the respective chat. I haven't found documentation how to actually unpin a message (also setting the pinned message to 0 doesn't help). But there will probably be a solution for that as well. Happy testing!

This will fix #57 completely (was already partially done before)

### CODE REVIEW

**Review by jgibbon — APPROVED** [review]
It seems to work really well, nice work!
None of the points I noticed and commented on are really bad and I think it could be merged, so I'll mark it as such.

One thing, though: 
I think the ListView is a bit too prominent in the background when looking at the message overlay.

On the one hand, it could be solved by lowering the opacity, but on the other hand this may be a good excuse for adding a bit of eye candy. So here comes my proposal for ChatPage to solve this "issue":
```
[...]
import QtGraphicalEffects 1.0
[...]
                Loader {
                    asynchronous: true
                    active: chatView.blurred
                    anchors.fill: chatView
                    sourceComponent: Component {
                        FastBlur {
                            source: chatView
                            radius: chatView.visible ? 0 : Theme.paddingLarge
                        }
                    }
                }

                SilicaListView {
                    id: chatView
[...]
                    visible: !blurred
                    property bool blurred: messageOverlayLoader.item
[...]
                }
[...]
```

**Comment by jgibbon** [comment]
Oh, two other things: 
- When the (pinned) overlay is open and the pinned message hidden, I'd expect the overlay to close, as well.
- For unpinning messages, this may work: https://core.telegram.org/tdlib/docs/classtd_1_1td__api_1_1unpin_chat_message.html

**Comment by Wunderfitz** [comment]
I wonder why I haven't found the unpin function... Anyway, I will add a few improvements including the unpin function before we merge this one here.

File: qml/components/MessageOverlayFlickable.qml [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                        overlayWebPagePreviewLoader.active = true;
+                    }
+                }
+            }
+        }
+    }
+
+    Rectangle {
+        id: messageContentBackground
+        color: (Theme.colorScheme === Theme.LightOnDark) ? Theme.darkSecondaryColor : Theme.lightSecondaryColor
```
[the remark this answers is not in the document]
Comment (jgibbon) [reply qml/components/MessageOverlayFlickable.qml:80]: Or `Theme.overlayBackgroundColor`?

File: qml/components/MessageListViewItem.qml [hunk 2]
```diff
@@ -101,6 +101,36 @@ ListItem {
...
+                        }
+                        if (page.chatGroupInformation.status["@type"] === "chatMemberStatusCreator") {
+                            console.log("Creator of this chat: Yes!");
+                            return true;
+                        }
+                        if (page.chatInformation.permissions.can_pin_messages) {
+                            console.log("All people can pin: Yes!");
+                            return true;
+                        }
+                        if (page.chatGroupInformation.status["@type"] === "chatMemberStatusAdministrator") {
```
[the remark this answers is not in the document]
Comment (Wunderfitz) [reply qml/components/MessageListViewItem.qml:120]: Seems so... Will integrate that as well!

File: qml/components/MessageOverlayFlickable.qml [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    property var overlayMessage;
+    property bool showHeader: true
+    readonly property var userInformation: tdLibWrapper.getUserInformation(overlayMessage.sender_user_id);
+    readonly property bool isOwnMessage: tdLibWrapper.getUserInformation().id === overlayMessage.sender_user_id;
+    readonly property string extraContentComponentName: (typeof overlayMessage.content !== "undefined" && typeof chatView.contentComponentNames[overlayMessage.content['@type']] !== "undefined" )
+                                                        ? chatView.contentComponentNames[overlayMessage.content['@type']] : ""
+    signal requestClose;
+
+    function getOriginalAuthor(forwardInformation, fontSize) {
+        if (forwardInformation.origin["@type"] === "messageForwardOriginChannel") {
```
[the remark this answers is not in the document]
Comment (Wunderfitz) [reply qml/components/MessageOverlayFlickable.qml:41]: Jep, that would be better... ;)

File: qml/components/MessageOverlayFlickable.qml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Wunderfitz) [reply qml/components/MessageOverlayFlickable.qml:80]: `Theme.overlayBackgroundColo` seems to work smoothly, but then I need to lower the opacity a bit. I'll go for that!

File: qml/pages/ChatPage.qml [hunk 4]
```diff
@@ -574,10 +594,18 @@ Page {
...
+                    messageOverlayLoader.overlayMessage = pinnedMessageItem.pinnedMessage;
+                    messageOverlayLoader.active = !messageOverlayLoader.active;
+                }
+            }
+
             Item {
                 id: chatViewItem
                 width: parent.width
-                height: parent.height - headerRow.height -  newMessageColumn.height - selectedMessagesActions.height
+                height: parent.height - headerRow.height - ( pinnedMessageItem.visible ? pinnedMessageItem.height : 0 ) - newMessageColumn.height - selectedMessagesActions.height
```
[the remark this answers is not in the document]
Comment (Wunderfitz) [reply qml/pages/ChatPage.qml:608]: Yep, good catch. Will change that!

**Comment by Wunderfitz** [comment]
Seems that I messed it up... Let's see what can be done here.

**Comment by Wunderfitz** [comment]
Closing and following up in #177

**Pull request closed without merging**